java/code/src/com/suse/manager/reactor/messaging/RegisterMinionEventMessageAction.java

```diff
@@ -218,61 +242,62 @@ private void registerMinion(String minionId, boolean isSaltSSH,
                     }
                 });
             }
-
-            return;
+            return true;
         }
+        return false;
+    }
 
-        if (duplicateMinionNamePresent(minionId)) {
-            return;
+    /**
+     * Complete the minion registration with information from grains
+     * @param minionId the minion id
+     * @param machineId the machine id that we are trying to register
+     * @param creator the optional User that created the minion
+     * @param saltSSHProxyId optional proxy id for saltssh in case it is used
+     * @param activationKeyOverride optional label of activation key to be applied to the system
+     * @param isSaltSSH true if a salt-ssh system is bootstrapped
+     */
+    public void finalizeMinionRegistration(String minionId,
+                                           String machineId,
+                                           Optional<User> creator,
+                                           Optional<Long> saltSSHProxyId,
+                                           Optional<String> activationKeyOverride,
+                                           boolean isSaltSSH) {
+
+        MinionServer server = migrateTraditionalOrGetNew(minionId,
```

Comment:
Please note that `MinionServer` and `Server` are two different classes. If possible, name this local variable `minion` to make it clear this is not an instance of `Server`